fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/api/registry/SpreadableBlockRegistry.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.registry;
+
+import com.google.common.collect.ImmutableMap;
+import org.jetbrains.annotations.ApiStatus;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.block.Block;
+import net.minecraft.block.BlockState;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.api.util.Block2ObjectMap;
+import net.fabricmc.fabric.impl.content.registry.SpreadableBlockRegistryImpl;
+
+/**
+ * Registries of blocks to block states, defining the spreadable block state to replace
+ * a bare block with when a particular type of spreadable block spreads to the bare block.
+ *
+ * <p>For example, to get the registry for mycelium block spread and then register
+ * a modded block pair where mycelium spreading to {@code MY_DIRT} generates {@code MY_MYCELIUM}:
+ *
+ * <pre>{@code
+ * SpreadableBlockRegistry.getInstance(SpreadableBlockRegistry.MYCELIUM)
+ *        .add(MyModBlocks.MY_DIRT, MyModBlocks.MY_MYCELIUM.getDefaultState());
+ * }</pre>
+ */
+@ApiStatus.NonExtendable
+public interface SpreadableBlockRegistry extends Block2ObjectMap<BlockState> {
+	/**
+	 * Registry ID for grass-type spreadable blocks.
+	 */
+	Identifier GRASS = new Identifier("grass");
+	/**
+	 * Registry ID for mycelium-type spreadable blocks.
+	 */
+	Identifier MYCELIUM = new Identifier("mycelium");
+
+	/**
+	 * Get the registry for a given registry ID, or create a new registry for the ID
+	 * if none currently exists.
+	 *
+	 * @param type The registry type Identifier for the desired spreadable block registry
+	 * @return The {@code SpreadableBlockRegistry} for the given ID
+	 */
+	static SpreadableBlockRegistry getInstance(Identifier type) {
+		return SpreadableBlockRegistryImpl.getOrCreateInstance(type);
+	}
+
+	/**
+	 * Gets an immutable copy of the spreadable block types map, which maps the
+	 * canonical or "primary" block of the type to the registry for the type.
+	 *
+	 * <p>For example, this could be used when implementing a custom block which
+	 * becomes whichever spreadable block there is the most of nearby.
+	 *
+	 * @return An immutable copy of the spreadable block types map
+	 */
+	static ImmutableMap<Block, Identifier> getSpreadableTypes() {
```

Comment:
Is there any reason to use indirect `Identifier`s as the values instead of the registry instance directly?

Reply:
I wonder if a RegistryKey is the soltuion here? or maybe leave that until we are forced into it?

Reply:
It's the id of a SpreadableBlockRegistry, not a vanilla registry entry like Block. In other words, just a key of this map:
https://github.com/FabricMC/fabric/blob/fee11564048c76a1273685b0496d559d5c1ce5a4/fabric-content-registries-v0/src/main/java/net/fabricmc/fabric/impl/content/registry/SpreadableBlockRegistryImpl.java#L41

The instances for each id are constant once they're created (and the spreadable block registries that are in the map have been created), so it'd be simpler to just have the registry itself in the map IMO.

Reply:
Ah yeah, sorry I didnt look at the full context. I think this would be a good idea to change.

Reply:
I concur with this.